File: leetcode-75/02-two-pointers/11-is-subsequence/solution.py
"""
LeetCode 392. Is Subsequence
https://leetcode.com/problems/is-subsequence/

Approach: two pointers, i over s and j over t. Advance j through t,
and whenever it matches the current character of s, advance i too.
s is a subsequence of t iff i reaches the end of s.

Time:  O(|t|) - single pass over t
Space: O(1) - two index pointers
"""


class Solution:
    def isSubsequence(self, s: str, t: str) -> bool:
        i = 0
        n = len(s)

        for ch in t:
            if i < n and ch == s[i]:
                i += 1

        return i == n


def assert_equal(actual, expected, label):
    if actual != expected:
        print(f"FAIL [{label}]: got {actual}, expected {expected}")
    else:
        print(f"PASS [{label}]")


if __name__ == "__main__":
    sol = Solution()
    print(sol.isSubsequence("abc", "ahbgdc"))  # True
    print(sol.isSubsequence("axc", "ahbgdc"))  # False
    print(sol.isSubsequence("", "ahbgdc"))     # True

    # Edge cases
    assert_equal(sol.isSubsequence("", ""), True, "both empty")
    assert_equal(sol.isSubsequence("a", ""), False, "non-empty s, empty t")
    assert_equal(sol.isSubsequence("abc", "abc"), True, "s equals t exactly")
    assert_equal(sol.isSubsequence("abcd", "abc"), False, "s longer than t")
    assert_equal(sol.isSubsequence("aaa", "aaaa"), True, "repeated char, enough occurrences")
    assert_equal(sol.isSubsequence("aaa", "aa"), False, "repeated char, one short")

File: leetcode-75/02-two-pointers/11-is-subsequence/solution.ts
/**
 * LeetCode 392. Is Subsequence
 * https://leetcode.com/problems/is-subsequence/
 *
 * Approach: two pointers, i over s and j over t. Advance j through t,
 * and whenever it matches the current character of s, advance i too.
 * s is a subsequence of t iff i reaches the end of s.
 *
 * Time:  O(|t|) - single pass over t
 * Space: O(1) - two index pointers
 */
function isSubsequence(s: string, t: string): boolean {
  let i = 0;

  for (let j = 0; j < t.length && i < s.length; j++) {
    if (t[j] === s[i]) {
      i++;
    }
  }

  return i === s.length;
}
function assertEqual(actual: unknown, expected: unknown, label: string): void {
  const a = JSON.stringify(actual);
  const e = JSON.stringify(expected);
  if (a !== e) {
    console.error(`FAIL [${label}]: got ${a}, expected ${e}`);
  } else {
    console.log(`PASS [${label}]`);
  }
}

if (require.main === module) {
  // Example usage:
  console.log(isSubsequence('abc', 'ahbgdc')); // true
  console.log(isSubsequence('axc', 'ahbgdc')); // false
  console.log(isSubsequence('', 'ahbgdc'));     // true

  // Edge cases
  assertEqual(isSubsequence('', ''), true, 'both empty');
  assertEqual(isSubsequence('a', ''), false, 'non-empty s, empty t');
  assertEqual(isSubsequence('abc', 'abc'), true, 's equals t exactly');
  assertEqual(isSubsequence('abcd', 'abc'), false, 's longer than t');
  assertEqual(isSubsequence('aaa', 'aaaa'), true, 'repeated char, enough occurrences');
  assertEqual(isSubsequence('aaa', 'aa'), false, 'repeated char, one short');
}

export { isSubsequence };
